src/lib/firmware/multiboot.c

```diff
@@ -0,0 +1,470 @@
+	if (((header->hdr.flags & ~supported_mandatory) & 0xFFFF) != 0x0)
+		die("multiboot: header has unsupported mandatory flags (0x%x), bailing.\n", header->hdr.flags & 0xFFFF);
+
+	if (!(header->hdr.flags & (1 << 16))) {
+		multiboot_parse_elf(s, img);
+		return;
+	}
+
+	s->kernel_offset = (uint32_t)((uintptr_t)header - (uintptr_t)img->mapping -
+		(header->lhdr.header_addr - header->lhdr.load_addr));
+	if (header->lhdr.load_end_addr) {
+		s->kernel_size = header->lhdr.load_end_addr - header->lhdr.load_addr;
+	} else {
+		s->kernel_size = img->size - s->kernel_offset;
+	}
+	s->kernel_load_addr = header->lhdr.load_addr;
+	s->kernel_entry_addr = header->lhdr.entry_addr;
+}
+
+// guest_to_host translates GPA to HVA.
+static uintptr_t
+guest_to_host(uint32_t guest_addr, struct multiboot_state *s)
+{
+	return s->guest_mem_base + guest_addr;
+}
+
+// multiboot_load_data loads data pointed to by from of size into current load
+// address in guest's memory. The load address is advanced taken align into
+// account.
+static uint32_t
+multiboot_load_data(struct multiboot_state *s, void *from, uint32_t size, unsigned align)
+{
+	uintptr_t to = guest_to_host(s->load_addr, s);
+	uint32_t loaded_at = s->load_addr;
+
+	if (from == NULL || size == 0)
+		return 0;
+
+	if ((s->load_addr >= s->guest_mem_size) ||
+			((s->load_addr + size) > s->guest_mem_size))
+		die("multiboot: %x+%x is beyond guest's memory\n", s->load_addr, size);
+
+	memcpy((void *)to, from, size);
+
+	s->load_addr = s->load_addr + size;
+	if (align)
+		s->load_addr = ROUND_UP(s->load_addr, align);
+
+	return loaded_at;
+}
+
+// multiboot_set_guest_state prepares guest registers/descriptors per Multiboot
+// specification.
+static uint64_t
+multiboot_set_guest_state(struct multiboot_state *s)
+{
+	xh_vcpu_reset(0);
+	xh_vm_set_register(0, VM_REG_GUEST_RAX, 0x2BADB002);
+	xh_vm_set_register(0, VM_REG_GUEST_RBX, s->mbi_addr);
+	xh_vm_set_register(0, VM_REG_GUEST_RIP, s->kernel_entry_addr);
+
+	xh_vm_set_desc(0, VM_REG_GUEST_CS, 0, 0xffffffff, 0xc09b);
+	xh_vm_set_desc(0, VM_REG_GUEST_DS, 0, 0xffffffff, 0xc093);
+	xh_vm_set_desc(0, VM_REG_GUEST_ES, 0, 0xffffffff, 0xc093);
+	xh_vm_set_desc(0, VM_REG_GUEST_FS, 0, 0xffffffff, 0xc093);
+	xh_vm_set_desc(0, VM_REG_GUEST_GS, 0, 0xffffffff, 0xc093);
+	xh_vm_set_desc(0, VM_REG_GUEST_SS, 0, 0xffffffff, 0xc093);
+
+	xh_vm_set_register(0, VM_REG_GUEST_CR0, 0x21);
+
+	return s->kernel_entry_addr;
+}
+
+// multiboot_add_cmdline add boot_file cmdline into combined command line
+// buffer.
+static uint32_t
+multiboot_add_cmdline(struct multiboot_state *s, char *boot_file, char *cmdline)
+{
+	uint32_t offset = s->cmdline_len;
+
+	unsigned len = (unsigned)strlen(boot_file) + 1;
+	if (cmdline != NULL)
+		len += (unsigned)strlen(cmdline) + 1;
+
+	s->cmdline = realloc(s->cmdline, s->cmdline_len + len);
+	if (cmdline != NULL)
+		snprintf(s->cmdline + s->cmdline_len, len, "%s %s", boot_file, cmdline);
+	else
+		snprintf(s->cmdline + s->cmdline_len, len, "%s", boot_file);
```

Comment:
Isn't it better to `die()` here if realloc fails rather than boot without the cmdline?  The latter behaviour would lead to rather non-deterministic boot behaviour under low memory conditions.